### PR TITLE
Execute generated user code in "dark-editor" canvas/experiment

### PR DESCRIPTION
**Pull request by StachuDotNet**
Changelog:

```
CanvasHack, experiment1
- `add Edtiro::parseAndExecuteExpr`
- execute generated user code within the editor
```

### CODE REVIEW

File: backend/testfiles/execution/internal.tests [hunk 1]
```diff
@@ -94,4 +94,9 @@ module UnlockedDB =
 
 module Ops =
   // getOpsForToplevel returns empty list for nonexistant TL
-  DarkInternal.getOpsForToplevel_v0 Test.getCanvasID 0 = []
\ No newline at end of file
+  DarkInternal.getOpsForToplevel_v0 Test.getCanvasID 0 = []
+
+
+module Eval =
+  // getOpsForToplevel returns empty list for nonexistant TL
```
Comment (StachuDotNet) [inline backend/testfiles/execution/internal.tests:101]: TODO: remove comment

File: backend/src/BackendOnlyStdLib/LibDarkInternal.fs [hunk 2]
```diff
@@ -798,4 +798,29 @@ human-readable data."
...
+        TResult(TVariable "a", TString)
+      description =
+        "Parses and executes arbitrary Dark code in the context of the current canvas."
+      fn =
+        internalFn (function
+          | state, _, [ DString code; DDict userInputs ] ->
+            uply {
+              let expr = Parser.RuntimeTypes.parseExprWithTypes [] code
+              let symtable = LibExecution.Interpreter.withGlobals state userInputs
+              let! evalResult = LibExecution.Interpreter.eval state symtable expr
```
Comment (StachuDotNet) [inline backend/src/BackendOnlyStdLib/LibDarkInternal.fs:820]: TODO: maybe use a 'cleaner' slate

File: backend/src/BackendOnlyStdLib/LibDarkInternal.fs [hunk 3]
```diff
@@ -798,4 +798,29 @@ human-readable data."
...
+
+
+
+    { name = fn "DarkInternal" "parseAndExecuteExpr" 0
+      typeParams = []
+      parameters =
+        [ Param.make "code" TString ""; Param.make "userInputs" (TDict TString) "" ]
+      returnType =
+        // TODO: improve type
+        TResult(TVariable "a", TString)
```
Comment (StachuDotNet) [inline backend/src/BackendOnlyStdLib/LibDarkInternal.fs:811]: TODO: declaration doesn't match reality. also, maybe there's a better structure to return than this

File: backend/src/BackendOnlyStdLib/BackendOnlyStdLib.fsproj [hunk 4]
```diff
@@ -12,6 +12,7 @@
   <ItemGroup>
     <ProjectReference Include="../Prelude/Prelude.fsproj" />
     <ProjectReference Include="../LibExecution/LibExecution.fsproj" />
+    <ProjectReference Include="../Parser/Parser.fsproj" />
```
Comment (pbiggar) [inline backend/src/BackendOnlyStdLib/BackendOnlyStdLib.fsproj:15]: Thinking out loud, perhaps we should have a BwdServerExperiment1 where we include these?

Comment (StachuDotNet) [reply backend/src/BackendOnlyStdLib/BackendOnlyStdLib.fsproj:15]: Hmm, what are "these" in this context?

File: backend/testfiles/execution/internal.tests [hunk 5]
```diff
@@ -98,5 +98,7 @@ module Ops =
 
 
 module Eval =
-  // getOpsForToplevel returns empty list for nonexistant TL
-  DarkInternal.parseAndExecuteExpr "1 + 2" Dict.empty_v0 = 3
\ No newline at end of file
+  DarkInternal.parseAndExecuteExpr "1 + 2" Dict.empty_v0 = 3
+  DarkInternal.parseAndExecuteExpr "a" { a = 3 } = 3
+  DarkInternal.parseAndExecuteExpr """let a = 3 in a + b""" { b = 2 } = 5
+  DarkInternal.parseAndExecuteExpr """(HttpClient.request "get" "https://darklang.com" [] Bytes.empty) |> Test.unwrap |> (fun response -> response.statusCode)""" Dict.empty = 200
```
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply backend/testfiles/execution/internal.tests:104]: fwiw does the same thing on example.com

File: backend/src/BackendOnlyStdLib/LibDarkInternal.fs (hunk 3 above)
Comment (StachuDotNet) [reply backend/src/BackendOnlyStdLib/LibDarkInternal.fs]: decided to just always return a string

File: backend/testfiles/execution/internal.tests (hunk 5 above)
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply backend/testfiles/execution/internal.tests]: not sure why these were failing in interpreter.tests, but equivalent usages pass in the actual dark-editor demo canvas so ignoring for now.

File: backend/src/BackendOnlyStdLib/LibDarkInternal.fs (hunk 2 above)
Comment (StachuDotNet) [reply backend/src/BackendOnlyStdLib/LibDarkInternal.fs:822]: is there risk in exposing the current `state` object? I don't think so but vaguely feels safer to potentially use a fresh 'ExecutionState' here. wdyt?

**Pull request merged**